File: src/main/java/dev/jlo/gamemodes/reward/RewardOutbox.java
package dev.jlo.gamemodes.reward;

import dev.jlo.gamemodes.persistence.SqlStatements;
import dev.jlo.gamemodes.persistence.SqliteMigrations;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.Clock;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class RewardOutbox {
    public static final long DEFAULT_LEASE_DURATION_MS = 60_000L;

    private final Connection connection;
    private final Clock clock;

    public RewardOutbox(Connection connection) {
        this(connection, Clock.systemUTC());
    }

    public RewardOutbox(Connection connection, Clock clock) {
        this.connection = connection;
        this.clock = clock;
        try {
            new SqliteMigrations(clock).apply(connection);
        } catch (Exception e) {
            throw new RuntimeException("Failed to apply reward outbox migrations", e);
        }
    }

    public synchronized boolean enqueue(String matchId, UUID playerId, String rewardType, long amount) {
        if (matchId == null || matchId.isBlank() || rewardType == null || rewardType.isBlank() || amount < 0) {
            throw new IllegalArgumentException();
        }
        try (PreparedStatement ps = connection.prepareStatement(SqlStatements.load("reward/insert-outbox.sql"))) {
            ps.setString(1, matchId);
            ps.setString(2, playerId.toString());
            ps.setString(3, rewardType);
            ps.setLong(4, amount);
            ps.setLong(5, clock.millis());
            return ps.executeUpdate() == 1;
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public synchronized List<Reward> claim() {
        return claim(100, DEFAULT_LEASE_DURATION_MS);
    }

    public synchronized List<Reward> claim(int limit) {
        return claim(limit, DEFAULT_LEASE_DURATION_MS);
    }

    public synchronized List<Reward> claim(int limit, long leaseDurationMs) {
        if (limit <= 0 || leaseDurationMs <= 0) {
            throw new IllegalArgumentException();
        }
        long now = clock.millis();
        long until = Math.addExact(now, leaseDurationMs);
        try {
            connection.setAutoCommit(false);
            List<String[]> keys = new ArrayList<>();
            try (PreparedStatement ps = connection.prepareStatement(SqlStatements.load("reward/select-claimable.sql"))) {
                ps.setLong(1, now);
                ps.setInt(2, limit);
                try (ResultSet rs = ps.executeQuery()) {
                    while (rs.next()) {
                        keys.add(new String[]{rs.getString(1), rs.getString(2), rs.getString(3)});
                    }
                }
            }
            if (!keys.isEmpty()) {
                try (PreparedStatement ps = connection.prepareStatement(SqlStatements.load("reward/update-claimed.sql"))) {
                    for (String[] key : keys) {
                        ps.setLong(1, now);
                        ps.setLong(2, until);
                        ps.setString(3, key[0]);
                        ps.setString(4, key[1]);
                        ps.setString(5, key[2]);
                        ps.setLong(6, now);
                        ps.addBatch();
                    }
                    ps.executeBatch();
                }
            }
            connection.commit();
            List<Reward> result = new ArrayList<>();
            for (String[] key : keys) {
                Reward reward = get(key[0], key[1], key[2]);
                if (reward != null) result.add(reward);
            }
            return result;
        } catch (Throwable t) {
            try {
                connection.rollback();
            } catch (SQLException rollbackFailure) {
                t.addSuppressed(rollbackFailure);
            }
            if (t instanceof RuntimeException runtimeException) throw runtimeException;
            if (t instanceof Error error) throw error;
            throw new RuntimeException(t);
        } finally {
            try {
                connection.setAutoCommit(true);
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        }
    }

    public synchronized List<Reward> pending() {
        return claim(100);
    }

    public synchronized List<Reward> pending(int limit) {
        return claim(limit);
    }

    public synchronized boolean markApplied(Reward reward) {
        try (PreparedStatement ps = connection.prepareStatement(SqlStatements.load("reward/update-applied.sql"))) {
            ps.setLong(1, clock.millis());
            ps.setString(2, reward.getMatchId());
            ps.setString(3, reward.getPlayerId().toString());
            ps.setString(4, reward.getType());
            return ps.executeUpdate() == 1;
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public synchronized int abortMatch(String matchId) {
        try (PreparedStatement ps = connection.prepareStatement(SqlStatements.load("reward/update-aborted.sql"))) {
            ps.setString(1, matchId);
            return ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private Reward get(String matchId, String playerId, String type) {
        try (PreparedStatement ps = connection.prepareStatement(SqlStatements.load("reward/select-outbox.sql"))) {
            ps.setString(1, matchId);
            ps.setString(2, playerId);
            ps.setString(3, type);
            try (ResultSet rs = ps.executeQuery()) {
                if (!rs.next()) return null;
                long lease = rs.getLong(4);
                Long leaseUntil = rs.wasNull() ? null : lease;
                return new Reward(matchId, UUID.fromString(playerId), type, rs.getLong(1),
                        RewardState.valueOf(rs.getString(2)), rs.getInt(3), leaseUntil);
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/dev/jlo/gamemodes/reward/RewardOutbox.kt
package dev.jlo.gamemodes.reward

import java.sql.Connection
import java.time.Clock
import java.util.UUID

enum class RewardState { PENDING, CLAIMED, APPLIED, ABORTED }
data class Reward(
    val matchId: String,
    val playerId: UUID,
    val type: String,
    val amount: Long,
    val state: RewardState,
    val attempts: Int = 0,
    val leaseUntilEpochMs: Long? = null
)

class RewardOutbox(private val connection: Connection, private val clock: Clock = Clock.systemUTC()) {
    init { dev.jlo.gamemodes.persistence.SqliteMigrations(clock).apply(connection) }

    @Synchronized fun enqueue(matchId: String, playerId: UUID, rewardType: String, amount: Long): Boolean {
        require(matchId.isNotBlank() && rewardType.isNotBlank() && amount >= 0)
        return connection.prepareStatement("INSERT INTO reward_outbox(match_id,player_id,reward_type,amount,state,created_at_epoch_ms) VALUES (?,?,?,?,'PENDING',?) ON CONFLICT(match_id,player_id,reward_type) DO NOTHING").use { ps ->
            ps.setString(1, matchId); ps.setString(2, playerId.toString()); ps.setString(3, rewardType)
            ps.setLong(4, amount); ps.setLong(5, clock.millis()); ps.executeUpdate() == 1
        }
    }

    /**
     * Atomically leases up to [limit] pending rewards. A claimed row is not
     * claimable again until its lease expires; each successful claim increments attempts.
     */
    @Synchronized fun claim(limit: Int = 100, leaseDurationMs: Long = DEFAULT_LEASE_DURATION_MS): List<Reward> {
        require(limit > 0 && leaseDurationMs > 0)
        val now = clock.millis()
        val until = Math.addExact(now, leaseDurationMs)
        connection.autoCommit = false
        try {
            val keys = connection.prepareStatement(
                "SELECT match_id,player_id,reward_type FROM reward_outbox " +
                    "WHERE state='PENDING' OR (state='CLAIMED' AND lease_until_epoch_ms<=?) " +
                    "ORDER BY created_at_epoch_ms,match_id,player_id,reward_type LIMIT ?"
            ).use { ps ->
                ps.setLong(1, now); ps.setInt(2, limit)
                ps.executeQuery().use { rs ->
                    buildList { while (rs.next()) add(Triple(rs.getString(1), rs.getString(2), rs.getString(3))) }
                }
            }
            if (keys.isNotEmpty()) {
                connection.prepareStatement(
                    "UPDATE reward_outbox SET state='CLAIMED', attempts=attempts+1, claimed_at_epoch_ms=?, lease_until_epoch_ms=? " +
                        "WHERE match_id=? AND player_id=? AND reward_type=? AND (state='PENDING' OR (state='CLAIMED' AND lease_until_epoch_ms<=?))"
                ).use { ps ->
                    keys.forEach { (matchId, playerId, type) ->
                        ps.setLong(1, now); ps.setLong(2, until); ps.setString(3, matchId); ps.setString(4, playerId)
                        ps.setString(5, type); ps.setLong(6, now); ps.addBatch()
                    }
                    ps.executeBatch()
                }
            }
            connection.commit()
            return keys.mapNotNull { get(it.first, it.second, it.third) }
        } catch (t: Throwable) {
            connection.rollback()
            throw t
        } finally {
            connection.autoCommit = true
        }
    }

    /** Compatibility alias: callers now receive leased, rather than read-only, rows. */
    @Synchronized fun pending(limit: Int = 100): List<Reward> = claim(limit)

    @Synchronized fun markApplied(reward: Reward): Boolean = connection.prepareStatement(
        "UPDATE reward_outbox SET state='APPLIED', applied_at_epoch_ms=?, lease_until_epoch_ms=NULL " +
            "WHERE match_id=? AND player_id=? AND reward_type=? AND state='CLAIMED'"
    ).use { ps ->
        ps.setLong(1, clock.millis()); ps.setString(2, reward.matchId); ps.setString(3, reward.playerId.toString())
        ps.setString(4, reward.type); ps.executeUpdate() == 1
    }

    @Synchronized fun abortMatch(matchId: String): Int = connection.prepareStatement(
        "UPDATE reward_outbox SET state='ABORTED', lease_until_epoch_ms=NULL WHERE match_id=? AND state IN ('PENDING','CLAIMED')"
    ).use { ps -> ps.setString(1, matchId); ps.executeUpdate() }

    private fun get(matchId: String, playerId: String, type: String): Reward? =
        connection.prepareStatement("SELECT amount,state,attempts,lease_until_epoch_ms FROM reward_outbox WHERE match_id=? AND player_id=? AND reward_type=?").use { ps ->
            ps.setString(1, matchId); ps.setString(2, playerId); ps.setString(3, type)
            ps.executeQuery().use { rs ->
                if (!rs.next()) null else Reward(matchId, UUID.fromString(playerId), type, rs.getLong(1), RewardState.valueOf(rs.getString(2)), rs.getInt(3), rs.getLong(4).let { if (rs.wasNull()) null else it })
            }
        }

    companion object { const val DEFAULT_LEASE_DURATION_MS: Long = 60_000 }
}
